### ISSUE TITLE
SQL generated by rebalance is too long for GenQuery (too many resources)

### ISSUE THREAD

**Issue by trel** [irods/irods#5110]
- [x] main
- [x] 4-2-stable
-------

### iRODS Version, OS and Version

seen in 4.2.7, confirmed still in 4-2-stable

### What did you try to do?

run `iadmin modresc replResc rebalance` on a resource tree with many resources under a replication resource.

In this case 216 resources....

### Expected behavior

everything to rebalance - no errors

### Observed behavior (including steps to reproduce, if applicable)

remote addresses: 127.0.1.1 ERROR: rcGeneralAdmin failed with error -306000 USER_STRLEN_TOOLONG 

From the log...
```
Sep  9 09:41:11 pid:16192 remote addresses: 127.0.0.1, ::1 ERROR: rstrcpy not enough space in dest, slen:2596, maxLen:2176, stacktrace:

Sep  9 09:41:11 pid:16192 NOTICE: _rsGenQuery: genQuery status = -306000
Sep  9 09:41:11 pid:16192 NOTICE: rsGenQuery: _rsGenQuery failed, status = -306000
Sep  9 09:41:11 pid:16192 remote addresses: 127.0.0.1, ::1 ERROR: iRODS Exception:
    file: /irods_git_repo/plugins/resources/replication/irods_repl_rebalance.cpp
    function: std::vector<ReplicaAndRescId> (anonymous namespace)::get_out_of_date_replicas_batch(rsComm_t *, const std::vector<leaf_bundle_t> &, const std::string &, const int)
    line: 224
    code: -306000 (USER_STRLEN_TOOLONG)
    message:
        rsGenQuery failed. genquery_inp contents:
maxRows: 500 continueInx: 0 rowOffset: 0
options: 0
selectInp.len: 3
    column: 401 DATA_ID
    options: 1
    column: 404 DATA_REPL_NUM
    options: 1
    column: 423 DATA_RESC_ID
    options: 1
sqlCondInp.len: 3
    column: 423 DATA_RESC_ID
    condition: = '10382' || = '10383' || = '10384' || = '10385' || = '10386' || = '10387' || = '10388' || = '10389' || = '10390' || = '10391' || = '10392' || = '10393' || = '10394' || = '10395' || = '10396' || = '10397' || = '10398' || = '10399' || = '10400' || = '10401' || = '10402' || = '10403' || = '10404' || = '10405' || = '10406' || = '10407' || = '10408' || = '10409' || = '10410' || = '10411' || = '10412' || = '10413' || = '10414' || = '10415' || = '10416' || = '10417' || = '10418' || = '10419' || = '10420' || = '10421' || = '10422' || = '10423' || = '10424' || = '10425' || = '10426' || = '10427' || = '10428' || = '10429' || = '10430' || = '10431' || = '10432' || = '10433' || = '10434' || = '10435' || = '10436' || = '10437' || = '10438' || = '10439' || = '10440' || = '10441' || = '10442' || = '10443' || = '10444' || = '10445' || = '10446' || = '10447' || = '10448' || = '10449' || = '10450' || = '10451' || = '10452' || = '10453' || = '10454' || = '10455' || = '10456' || = '10457' || = '10458' || = '10459' || = '10460' || = '10461' || = '10462' || = '10463' || = '10464' || = '10465' || = '10466' || = '10467' || = '10468' || = '10469' || = '10470' || = '10471' || = '10472' || = '10473' || = '10474' || = '10475' || = '10476' || = '10477' || = '10478' || = '10479' || = '10480' || = '10481' || = '10282' || = '10283' || = '10284' || = '10285' || = '10286' || = '10287' || = '10288' || = '10289' || = '10290' || = '10291' || = '10292' || = '10293' || = '10294' || = '10295' || = '10296' || = '10297' || = '10298' || = '10299' || = '10300' || = '10301' || = '10302' || = '10303' || = '10304' || = '10305' || = '10306' || = '10307' || = '10308' || = '10309' || = '10310' || = '10311' || = '10312' || = '10313' || = '10314' || = '10315' || = '10316' || = '10317' || = '10318' || = '10319' || = '10320' || = '10321' || = '10322' || = '10323' || = '10324' || = '10325' || = '10326' || = '10327' || = '10328' || = '10329' || = '10330' || = '10331' || = '10332' || = '10333' || = '10334' || = '10335' || = '10336' || = '10337' || = '10338' || = '10339' || = '10340' || = '10341' || = '10342' || = '10343' || = '10344' || = '10345' || = '10346' || = '10347' || = '10348' || = '10349' || = '10350' || = '10351' || = '10352' || = '10353' || = '10354' || = '10355' || = '10356' || = '10357' || = '10358' || = '10359' || = '10360' || = '10361' || = '10362' || = '10363' || = '10364' || = '10365' || = '10366' || = '10367' || = '10368' || = '10369' || = '10370' || = '10371' || = '10372' || = '10373' || = '10374' || = '10375' || = '10376' || = '10377' || = '10378' || = '10379' || = '10380' || = '10381'
    column: 413 DATA_REPL_STATUS
    condition: = '0'
    column: 420 DATA_MODIFY_TIME
    condition: <= '01599658871'


possible iquest [select DATA_ID, DATA_REPL_NUM, DATA_RESC_ID where DATA_RESC_ID = '10382' || = '10383' || = '10384' || = '10385' || = '10386' || = '10387' || = '10388' || = '10389' || = '10390' || = '10391' || = '10392' || = '10393' || = '10394' || = '10395' || = '10396' || = '10397' || = '10398' || = '10399' || = '10400' || = '10401' || = '10402' || = '10403' || = '10404' || = '10405' || = '10406' || = '10407' || = '10408' || = '10409' || = '10410' || = '10411' || = '10412' || = '10413' || = '10414' || = '10415' || = '10416' || = '10417' || = '10418' || = '10419' || = '10420' || = '10421' || = '10422' || = '10423' || = '10424' || = '10425' || = '10426' || = '10427' || = '10428' || = '10429' || = '10430' || = '10431' || = '10432' || = '10433' || = '10434' || = '10435' || = '10436' || = '10437' || = '10438' || = '10439' || = '10440' || = '10441' || = '10442' || = '10443' || = '10444' || = '10445' || = '10446' || = '10447' || = '10448' || = '10449' || = '10450' || = '10451' || = '10452' || = '10453' || = '10454' || = '10455' || = '10456' || = '10457' || = '10458' || = '10459' || = '10460' || = '10461' || = '10462' || = '10463' || = '10464' || = '10465' || = '10466' || = '10467' || = '10468' || = '10469' || = '10470' || = '10471' || = '10472' || = '10473' || = '10474' || = '10475' || = '10476' || = '10477' || = '10478' || = '10479' || = '10480' || = '10481' || = '10282' || = '10283' || = '10284' || = '10285' || = '10286' || = '10287' || = '10288' || = '10289' || = '10290' || = '10291' || = '10292' || = '10293' || = '10294' || = '10295' || = '10296' || = '10297' || = '10298' || = '10299' || = '10300' || = '10301' || = '10302' || = '10303' || = '10304' || = '10305' || = '10306' || = '10307' || = '10308' || = '10309' || = '10310' || = '10311' || = '10312' || = '10313' || = '10314' || = '10315' || = '10316' || = '10317' || = '10318' || = '10319' || = '10320' || = '10321' || = '10322' || = '10323' || = '10324' || = '10325' || = '10326' || = '10327' || = '10328' || = '10329' || = '10330' || = '10331' || = '10332' || = '10333' || = '10334' || = '10335' || = '10336' || = '10337' || = '10338' || = '10339' || = '10340' || = '10341' || = '10342' || = '10343' || = '10344' || = '10345' || = '10346' || = '10347' || = '10348' || = '10349' || = '10350' || = '10351' || = '10352' || = '10353' || = '10354' || = '10355' || = '10356' || = '10357' || = '10358' || = '10359' || = '10360' || = '10361' || = '10362' || = '10363' || = '10364' || = '10365' || = '10366' || = '10367' || = '10368' || = '10369' || = '10370' || = '10371' || = '10372' || = '10373' || = '10374' || = '10375' || = '10376' || = '10377' || = '10378' || = '10379' || = '10380' || = '10381' and DATA_REPL_STATUS = '0' and DATA_MODIFY_TIME <= '01599658871']
        failed to rebalance resource: [-]	/irods_git_repo/server/api/src/rsGeneralAdmin.cpp:903:int _rsGeneralAdmin(rsComm_t *, generalAdminInp_t *) :  status [USER_STRLEN_TOOLONG]  errno [] -- message [failed to rebalance resource]
```

Proposed alternative format:
```
iquest "select DATA_ID, DATA_REPL_NUM, DATA_RESC_ID where DATA_RESC_ID in [ '818898' ,'818899' ,'1296443' , '1296444' , '1300727' , '1300744', '1404346', '1404347', '1404348', '1404349', '1404350', '1404371', '1404372', '3319742', '3319746', '3319762', '3319763', '3319764', '3319779', '3319780', '3319781', '3319782', '3319783', '3319784', '3319787', '3319788', '3319790', '3319791', '3319792', '3319794', '3319795', '3319796', '7140155', '7140156', '7140175', '7140190', '7140191', '14685333', '14808449', '17010606', '17010612', '17010621', '17010624', '18054711', '18054754', '18054765', '18058943', '18058944', '18058962', '18092280', '18092326', '18092341', '18095346', '18095393', '18095436', '18129894', '18129933', '18129935', '18160003', '18160004', '50795684', '50795685', '50795695', '50795696', '50795706', '50795708', '50795713', '50795714', '50795718', '50795719', '50795720', '50795721', '218587968', '223329509', '234876563', '234876688', '234876719', '234876799', '234876836', '234876837', '242970662', '242970664', '242970665', '242971148', '242971154', '242971167', '283683602', '284408543', '287997354', '289824564', '289824565', '289824566', '289825015', '289825016', '289825017', '290380734', '290380736', '290380737', '290380752', '290380753', '1296441', '1296442', '1300716', '1300740', '1404355', '1404373', '1404377', '1404378', '1404379', '1404380', '1404381', '3319624', '3319641', '3319642', '3319643', '3319645', '3319662', '3319665', '3319666', '3319668', '3319670', '3319671', '3319672', '3319673', '3319674', '3319687', '3319689', '3319690', '3319691', '3319692', '7140129', '7140154', '7140173', '7140174', '14685323', '14808418', '17010516', '17010518', '17010520', '17010522', '17010526', '18308889', '18375096', '18375477', '18375998', '18376124', '18376189', '18376257', '18376358', '20004858', '20259075', '20259170', '20259197', '20259252', '20259253', '20259272', '20283350', '20283857', '20283998', '20284118', '20284305', '20284319', '20284550', '20284622', '20284718', '20284719', '20284744', '20284746', '20284820', '20320328', '22540222', '22540238', '22540252', '22753334', '22754383', '22754584', '22754766', '23354046', '25676244', '26059804', '36450933', '50795689', '50795690', '50795691', '50795697', '50795700', '50795701', '50795712', '50795715', '50795716', '50795717', '50795725', '50795726', '158085796', '158085797', '158085904', '158085905', '158085906', '242878930', '242878940', '242879010', '242879048', '242879049', '242879050', '290049690', '290049691', '290050358', '290050370', '290050374', '290050379', '291878826', '291878842', '291878859', '291878874', '291878914', '291878929'] and DATA_REPL_STATUS = '0' and DATA_MODIFY_TIME <= '01599654296'"
```

rebalance code of interest
https://github.com/irods/irods/blob/93984173cde6c5e721a9a828dbc1b31ffe38f415/plugins/resources/replication/irods_repl_rebalance.cpp#L175-L201

- have confirmed that GenQuery can use the `IN()` syntax...

**Comment by trel** [comment]
inspiration for the earlier #5073

**Comment by kript** [comment]
Has this been fixed by #4731 ?

**Comment by kript** [comment]
Oh, bother. Do you have a timeline for it being fixed then, he said, hopefully?

**Comment by trel** [comment]
we can target 4.2.12... so it stays in the 4.2.x family.

4.2.11 is basically done and being tested now.

**Comment by trel** [comment]
a candidate...  https://github.com/irods/irods/pull/6028

**Comment by trel** [comment]
The solution here only 'kicks the can down the road'... by using a more compressed SQL syntax.    

We're going to save 5 characters-per-resource... so at 300x5=1500... that's 1500 characters of RESC_IDs... or maybe... an additional 200 resources?

So rather than hitting this limit at ~300 resources, it might get pushed out to ~500 or so?

We'll still have a hard cap on the string length itself.

**Comment by kript** [comment]
We're planning to upgrade in the next couple of months, and it would be really good to have this functionality available when we do, as we currently maintain a manual script that does what the rebalance is supposed to do.

How confident are you around the ~300 resources number? We have been retiring a lot of resources recently and our biggest tree now has 'only' 258 resources, so it would be great if we could resume using that even in 4.2.7...

Of course, I might be counting the wrong things;

```
$ ilsresc root | wc -l
277
$ ilsresc root | grep -c unix
258
$ ilsresc | wc -l
348
```

**Comment by kript** [comment]
Oh, and is it in 4.3.0, and if not, can we get it added to 4.3.1 candidate, unless that has a different fix and so doesn't need it (he said, hopefully)?

**Comment by trel** [comment]
The syntax compression from the fix for this issue was included in 4.2.11 and 4.3.0.

With your initially reported '216' resources, the SQL hit the string limit with `slen:2596, maxLen:2176`.

With the compressed syntax, I'd think the SQL would save 216x5~=1000 characters... so the total SQL string length from Sept 2020 would have been around 1500 characters, and would fit within the `maxLen:2176`.

However, you say you now have 258 resources... so 216/258 = 1500/x... solve for x... ~1800 ... so... yes?  Upon upgrade, 4.2.11 should support your 258 resources under a single rebalance.  And perform much faster because of #4731 in 4.2.8.
